### PR TITLE
MEN-5475: meta-mender-demo: Add demo functionality for mender-gateway

### PR DESCRIPTION
**Pull request by lluiscampos**
Main commit:

    MEN-5475: meta-mender-demo: Add demo functionality for mender-gateway
    
    Requires the user to extend SRC_URI of mender-gateway with the location
    of the examples tarball:
    ```
    SRC_URI_pn-mender-gateway_append = " file:///<download-path>/mender-gateway-examples-1.0.0.tar"
    ```
    
    Download the tarball from https://docs.mender.io/downloads
    
    Changelog: Commit

### CODE REVIEW

**Comment by lluiscampos** [comment]
These go together:
* https://github.com/mendersoftware/mender-gateway/pull/17
* https://github.com/mendersoftware/mender-qa/pull/572
* https://github.com/mendersoftware/meta-mender/pull/1613
* https://github.com/mendersoftware/mender-dist-packages/pull/216

**Review by merlin-northern — APPROVED** [review]
pre-approving, not an expert. just some comments and curious questions.

File: meta-mender-demo/mender-commercial/mender-gateway/mender-gateway_%.bbappend [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+FILES_${PN}-doc_append = " \
+    ${docdir}/mender-gateway/examples \
+    ${sysconfdir}/mender/ \
+"
+
+DEPENDS_append = " rsync-native"
+
+def examples_dir_from_s_dir(d, s):
+    return s.replace("mender-gateway-", "mender-gateway-examples-")
+
```
Comment (merlin-northern) [inline meta-mender-demo/mender-commercial/mender-gateway/mender-gateway_%.bbappend:11]: yocto magician!

File: meta-mender-demo/mender-commercial/mender-gateway/mender-gateway_%.bbappend [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+EXAMPLES = "${@examples_dir_from_s_dir(d, '${S}')}"
+
+do_install_append() {
+    rsync --links --recursive ${EXAMPLES}/* ${D}
```
[the remark this answers is not in the document]
Comment (TheYoctoJester) [reply meta-mender-demo/mender-commercial/mender-gateway/mender-gateway_%.bbappend:15]: @merlin-northern yes, this is almost certainly incorrect. Especially due to `rsync` not being part of the standard `HOSTTOOLS`, this means it creates a hidden dependency on rsync being installed on the host machine. Even if it is verbose, the usual best practise is to use `cp -R --no-dereference --preserve=mode,links -v SOURCE DESTINATION` and install links manually as needed.

[the remark this answers is not in the document]
Comment (lluiscampos) [reply meta-mender-demo/mender-commercial/mender-gateway/mender-gateway_%.bbappend:15]: @TheYoctoJester Thanks. We were already doing it [here](https://github.com/mendersoftware/meta-mender/blob/9281533cd2fd8a542c36655eaa8f7138c7f70459/tests/meta-mender-ci/recipes-testing/mender-coverage-client/mender-client_%25.bbappend#L31) and [here](https://github.com/mendersoftware/meta-mender/blob/9281533cd2fd8a542c36655eaa8f7138c7f70459/tests/meta-mender-ci/recipes-testing/mender-artifact-coverage/mender-artifact_%25.bbappend#L27) so I thought it was okey. But I'll amend following your `cp` suggestion

> Especially due to `rsync` not being part of the standard `HOSTTOOLS`, this means it creates a hidden dependency on rsync being installed on the host machine

The dependency is added few lines above with ` DEPENDS_append = " rsync-native"`

[the remark this answers is not in the document]
Comment (TheYoctoJester) [reply meta-mender-demo/mender-commercial/mender-gateway/mender-gateway_%.bbappend:15]: @lluiscampos Okay, I did not notice that, sorry. I don't think rewriting things is necessary or a good idea, but for new code I would like to go with the canonical form if possible.

**Review by oleorhagen — APPROVED** [review]

**Pull request merged**